### PR TITLE
Add encoding for FixedSizeBinary and FixedSizeList

### LINKED ISSUE #15: Fixed size binary encoding
# Problem Statement

Fixed size binary encoding is very useful when the users know that each blob has the same size, i.e., storing prepared Tensors. It allows fast point query without lookups on the offsets. 

Additionally, since the data is aligned on the disk/cloud storage, it allows the "dataset loader" to read a batch of tensors (either in `BCHW` or `BCWH` form) without further transposing them in memory again.

### PR DESCRIPTION
**Pull request by eddyxu**
Closes #15

### CODE REVIEW

**Comment by eddyxu** [comment]
One thing I am not very sure is whether we should merge this code with plain encoding. Conceptually they are very similar.

File: cpp/src/lance/arrow/type.cc [hunk 1]
```diff
@@ -181,6 +182,41 @@ ::arrow::Result<std::shared_ptr<::arrow::DataType>> FromLogicalType(
       "FromLogicalType: logical_type \"{}\" is not supported yet", logical_type.to_string()));
 }
 
+::arrow::Result<std::shared_ptr<::arrow::ArrayBuilder>> GetArrayBuilder(
```
Comment (changhiskhan) [inline cpp/src/lance/arrow/type.cc:185]: This is only for numeric types?

File: cpp/src/lance/encodings/fixed_size_binary.cc [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+::arrow::Result<int64_t> FixedSizeBinaryEncoder::Write(const std::shared_ptr<::arrow::Array>& arr) {
+  assert(::arrow::is_fixed_size_binary(arr->type_id()) ||
+         lance::arrow::is_fixed_size_list(arr->type()));
+
+  ARROW_ASSIGN_OR_RAISE(auto values_position, out_->Tell());
+  if (::arrow::is_fixed_size_binary(arr->type_id())) {
+    auto a = std::static_pointer_cast<::arrow::FixedSizeBinaryArray>(arr);
+    ARROW_RETURN_NOT_OK(out_->Write(a->raw_values(), a->length() * a->byte_width()));
+  } else if (lance::arrow::is_fixed_size_list(arr->type())) {
+    auto list_arr = std::dynamic_pointer_cast<::arrow::FixedSizeListArray>(arr);
```
Comment (changhiskhan) [inline cpp/src/lance/encodings/fixed_size_binary.cc:40]: why use a dynamic cast here? i thought we already first check that it's a FixedSizeListArray?

File: cpp/src/lance/encodings/fixed_size_binary.cc [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  return values_position;
+}
+
+std::string FixedSizeBinaryEncoder::ToString() const { return "FixedSizeBinaryEncoder"; }
+
+::arrow::Result<std::shared_ptr<::arrow::Array>> FixedSizedBinaryDecoder::ToFixedSizeBinaryArray(
+    int32_t start, int32_t length) const {
+  auto bytes = type_->byte_width();
```
Comment (changhiskhan) [inline cpp/src/lance/encodings/fixed_size_binary.cc:54]: nit: I'd probably call this nbytes or byte_width?

File: cpp/src/lance/encodings/fixed_size_binary.cc [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    return ToFixedSizeBinaryArray(start, length.value());
+  }
+  return ::arrow::Status::Invalid("Invalid data type: {}\n", type_->ToString());
+}
+
+::arrow::Result<std::shared_ptr<::arrow::Array>> FixedSizedBinaryDecoder::Take(
+    std::shared_ptr<::arrow::Int32Array> indices) const {
+  std::shared_ptr<::arrow::ArrayBuilder> builder;
+  if (lance::arrow::is_fixed_size_list(type_)) {
+    auto list_type = std::dynamic_pointer_cast<::arrow::FixedSizeListType>(type_);
```
Comment (changhiskhan) [inline cpp/src/lance/encodings/fixed_size_binary.cc:98]: same question here wrt dynamic cast

File: cpp/src/lance/encodings/fixed_size_binary.cc [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    ARROW_ASSIGN_OR_RAISE(auto value_builder,
+                          ::lance::arrow::GetArrayBuilder(list_type->value_type()));
+    builder = std::make_shared<::arrow::FixedSizeListBuilder>(pool_, value_builder, type_);
+  } else if (::arrow::is_fixed_size_binary(type_->id())) {
+    builder = std::make_shared<::arrow::FixedSizeBinaryBuilder>(type_);
+  } else {
+    return ::arrow::Status::Invalid("FixedSizeBuilderDecoder::Take: Invalid data type: ", type_);
+  }
+
+  // TODO: Use thread pool
```
Comment (changhiskhan) [inline cpp/src/lance/encodings/fixed_size_binary.cc:108]: Should this be done for all decoders? Or just ones where we may end up reading large amounts of data for each scalar?

File: cpp/src/lance/encodings/fixed_size_binary.cc [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+  return values_position;
+}
+
+std::string FixedSizeBinaryEncoder::ToString() const { return "FixedSizeBinaryEncoder"; }
+
+::arrow::Result<std::shared_ptr<::arrow::Array>> FixedSizedBinaryDecoder::ToFixedSizeBinaryArray(
+    int32_t start, int32_t length) const {
+  auto bytes = type_->byte_width();
+  ARROW_ASSIGN_OR_RAISE(auto buf, infile_->ReadAt(position_ + start * bytes, length * bytes));
```
Comment (changhiskhan) [inline cpp/src/lance/encodings/fixed_size_binary.cc:55]: So for fixed size binary it's just a plain encoder/decoder using fixed width data (e.g., any of the numeric dtypes) right? How come we don't just call plain_decoder like you do with the fixed size list handling?

File: cpp/src/lance/encodings/fixed_size_binary.cc [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+#include <arrow/builder.h>
+
+#include "lance/arrow/type.h"
+#include "lance/encodings/encoder.h"
+#include "lance/encodings/plain.h"
+
+namespace lance::encodings {
+
+FixedSizeBinaryEncoder::FixedSizeBinaryEncoder(
```
Comment (changhiskhan) [inline cpp/src/lance/encodings/fixed_size_binary.cc:25]: This is called FixedSizeBinaryEncoder/Decoder but it handles both FixedSizeBinary and FixedSizeList. Would it be better to call it FixedSizeEncoder/Decoder or something like that?

**Review by changhiskhan — COMMENTED** [review]
I see what you mean about this vs plain codec. Your understanding of the code is better than mine, but based on my reading, I would probably prefer to have a separate abstraction for this. The plain codec can be kept simple and be made reusable.

Is there any particular reason you want to merge the two? Other just reducing LOC?

File: cpp/src/lance/encodings/fixed_size_binary.cc (hunk 2 above)
Comment (eddyxu) [reply cpp/src/lance/encodings/fixed_size_binary.cc:40]: So I was thinking of being consistent to use a dynamic cast to cast base class to child class. `dynamic_cast` is safer w.r.t. to virtual tables and etc.

I should use the same in line 37 to use dynamic_cast tho.

File: cpp/src/lance/encodings/fixed_size_binary.cc (hunk 7 above)
Comment (eddyxu) [reply cpp/src/lance/encodings/fixed_size_binary.cc:25]: Good point. So this encoding is in contrast to `VarBinary`, (var-sized binary). 
Maybe this is another argument to support merging with PlainEncoding

File: cpp/src/lance/encodings/fixed_size_binary.cc (hunk 6 above)
Comment (eddyxu) [reply cpp/src/lance/encodings/fixed_size_binary.cc:55]: Yes , we can definitely do it.

File: cpp/src/lance/encodings/fixed_size_binary.cc (hunk 5 above)
Comment (eddyxu) [reply cpp/src/lance/encodings/fixed_size_binary.cc:108]: This should be done if we choose to pick "scalar" individually. Usually it works better for the case where scalar is big, i.e., large binary, tensors, images and etc.

Other cases, i.e., for numeric values, it is faster to read a block and then filter them. (i.e., 64KB = 8000 int64 values in one read)

**Comment by eddyxu** [comment]
> Is there any particular reason you want to merge the two? Other just reducing LOC?

So conceptually, plain encoder is a special case of a fixed-size binary encoding with `byte_width = size of(primitive type)`. They supposedly share the same code path, and future same optimizations as well.

It use less "encoding enum" too, so in other language re-implementation, one less encoder to implement as well.

Parquet uses the plain encoding for all of them tho.

https://parquet.apache.org/docs/file-format/data-pages/encodings/

File: cpp/src/lance/arrow/type.cc (hunk 1 above)
Comment (eddyxu) [reply cpp/src/lance/arrow/type.cc:194]: Updated to support more.

**Review by changhiskhan — APPROVED** [review]
Main remaining item: do we also need to add boolean and timestamp since sophon is going to need it?
One thing with timestamps is that it's a parametric type so you'll need the unit. The default conversion from pandas date range is timestamp[ns].
Or is that a separate thing?

Looking at the code, you're right that merging it makes more sense. Code is simpler this way.

File: cpp/src/lance/arrow/type.cc [hunk 8]
```diff
@@ -181,6 +182,57 @@ ::arrow::Result<std::shared_ptr<::arrow::DataType>> FromLogicalType(
...
+::arrow::Result<std::shared_ptr<::arrow::ArrayBuilder>> GetFixedSizeListArrayBuilder(
+    const std::shared_ptr<::arrow::DataType>& dtype, ::arrow::MemoryPool* pool) {
+  assert(dtype->id() == ::arrow::Type::FIXED_SIZE_LIST);
+  auto list_type = std::dynamic_pointer_cast<::arrow::FixedSizeListType>(dtype);
+  ARROW_ASSIGN_OR_RAISE(auto value_builder,
+                        ::lance::arrow::GetArrayBuilder(list_type->value_type(), pool));
+  return std::make_shared<::arrow::FixedSizeListBuilder>(pool, value_builder, list_type);
+}
+
+::arrow::Result<std::shared_ptr<::arrow::ArrayBuilder>> GetArrayBuilder(
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply cpp/src/lance/arrow/type.cc:194]: Sure, will do

File: cpp/src/lance/encodings/plain.cc [hunk 9]
```diff
@@ -110,33 +123,34 @@ class PlainDecoderImpl : public Decoder {
...
       return ::arrow::Status::IndexError(
-          fmt::format("PlainDecoder::ToArray: out of range: start={}, length={}, page_length={}\n",
+          fmt::format("{}::ToArray: out of range: start={}, length={}, page_length={}\n",
+                      ToString(),
                       start,
                       length.value(),
                       length_));
     }
-    auto bytes = std::max(1, ::arrow::bit_width(type_->id()) / 8);
+    auto nbyes = type_->byte_width();
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply cpp/src/lance/encodings/plain.cc:139]: Fixed

**Pull request merged**